### ISSUE TITLE
panic when appending to sequence - maximum depth exceeded

### ISSUE THREAD

**Issue by iancoleman** [maidsafe/sn_api#674]
Appending to a sequence more than ~40 times causes a panic in the cli and an exit with status code 101

To reproduce

* get cli version
    * `safe --version` = `sn_cli 0.17.0` commit `67b746f6`
* start baby fleming network
    * `safe node run-baby-fleming`
* create account and log in
    * `safe auth start; safe auth create --test-coins; safe auth unlock --self-auth`
* create a sequence
    * `safe seq store firstval`
    * `export SEQURL=<the safe://url from the response>`
* using the sequence url, append 50 times
    * `safe seq append myval $SEQURL` or in one line:
    * `for i in {1..50}; do echo "appending $i of 50"; safe seq append myval $SEQURL; done`
* observe the cli panics after about 40 appends

This happens regardless of the size of the content being appended. I've created sequences up to 1.5 MiB but the size is not the cause of the panic (1.5 MiB from approx 32 KiB per append x 48 appends). The example above also panics despite storing small values.

I've seen panics after 41 appends, 48 appends, 39 appends, 35 appends, 51 appends which is strangely irregular.

The panic report from /tmp/report-<id>.toml is below, and a convient link to [crdts-4.3.0/src/lseq/ident.rs#L215](https://github.com/rust-crdt/rust-crdt/blob/4.3.0/src/lseq/ident.rs#L215)

```text
name = 'sn_cli'
operating_system = 'unix:Ubuntu'
crate_version = '0.17.0'
explanation = '''
Panic occurred in file '/home/ian/.cargo/registry/src/github.com-1ecc6299db9ec823/crdts-4.3.0/src/lseq/ident.rs' at line 215
'''
cause = 'maximum depth exceeded'
method = 'Panic'
backtrace = '''

   0: 0x7f90c3fd40a5 - crdts::lseq::ident::IdentGen<A>::alloc::h2cb7c8c0203b42a0
   1: 0x7f90c3fcb84b - crdts::lseq::LSeq<T,A>::append::h3d0110e3a080870e
   2: 0x7f90c3fd6ceb - sn_data_types::sequence::seq_crdt::SequenceCrdt<A,P>::create_append_op::h5b7d7b3243b804f5
   3: 0x7f90c3fdd0ed - sn_data_types::sequence::Data::create_unsigned_append_op::hbdaccdd4185dcf33
   4: 0x7f90c37b1dd9 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h58722b2859eaa6be
   5: 0x7f90c37fe57a - safe::cli::run_with::{{closure}}::h1b0f4823c2d116bc
   6: 0x7f90c37af653 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h40bee3cea4652556
   7: 0x7f90c38ea7c2 - tokio::runtime::thread_pool::ThreadPool::block_on::he3e6bc87cf2b03b1
   8: 0x7f90c388ed3a - tokio::runtime::context::enter::h6074c651c7f1d794
   9: 0x7f90c3a23dd8 - safe::main::hc02618c5d353558f
  10: 0x7f90c39a4633 - std::sys_common::backtrace::__rust_begin_short_backtrace::h33d90f38141ab49c
  11: 0x7f90c39a4869 - std::rt::lang_start::{{closure}}::h0876ddd9157edd35
  12: 0x7f90c444f2d9 - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::hefea566cf07ef0a6
                at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/ops/function.rs:259
                 - std::panicking::try::do_call::h5f7362256436cfed
                at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panicking.rs:381
                 - std::panicking::try::h1703f234309e3e9c
                at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panicking.rs:345
                 - std::panic::catch_unwind::h54d69a771c3636a5
                at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panic.rs:396
                 - std::rt::lang_start_internal::h53bf2c2c6c9e97dc
                at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/rt.rs:51
  13: 0x7f90c3a23f02 - main'''
```

**Comment by iancoleman** [comment]
This panic is caused by the choice of LSEQ_BOUNDARY in sn_data_types:

https://github.com/maidsafe/sn_data_types/blob/b0796d1a4baf9f53d68e07425eb20791651da382/src/sequence/seq_crdt.rs#L26-L31

The rust-crdt repo has relevant comments:

https://github.com/rust-crdt/rust-crdt/blob/de13805fea33dc1c6eea0560bff5bbd1cbe66648/src/lseq/ident.rs#L6-L9

I'm not sure what the appropriate trade-off here is, but boundary=1 is far too restrictive, only allowing about 40 append operations before panicking at the limit here [L215:base_bits](https://github.com/rust-crdt/rust-crdt/blob/4.3.0/src/lseq/ident.rs#L215). When I changed boundary to 500, base_bits did not increase as fast so a lot more appends could be done (I exited the loop with no panic after 107 appends and base_bits=22; appending gets slower for each call to append which is a bit of a red flag but getting out of scope for this issue).

Any thoughts on how to pick these values? Can they be dynamic / smart or must they be const?

**Comment by bochaco** [comment]
This is very interesting finding @iancoleman , I'll try to spend some time next week to analyse it in detail, I'll have to refresh my memory on all this. As you say and from what I remember there is a trade off there with those constants but still I'm surprised that a height limit is being reached, since the arity is quite high as well.

**Comment by iancoleman** [comment]
I've tried a test in rust-crdt calling append 1000 times (a modification of [this test](https://github.com/rust-crdt/rust-crdt/blob/492705097c2ed198a400eb4ed1a941e34bccae1f/test/lseq.rs#L59-L72)) using tree_base=10 and boundary=1. It works fine, base_bits was always between 10-12. So the blowout on base_bits is not caused by rust-crdt. I think modifying the tree_base and boundary values are not the right way to fix it, they can stay at 10 and 1.

The next closest code to rust-crdt is in sn_data_types. I added a test calling append 1000 times in sn_data_types sequence/mod.rs (a modification of [this test](https://github.com/maidsafe/sn_data_types/blob/b0796d1a4baf9f53d68e07425eb20791651da382/src/sequence/mod.rs#L502-L557)) and this also did not panic. base_bits was also between 10-12, so this a positive sign with it matching the underlying crdts crate.

The next closest code is to that is sn_client (I think), and I've tried modifying [this test](https://github.com/maidsafe/sn_client/blob/c15483093c5d53bbfedc6370ec8fff2ae191f182/src/client/sequence_apis.rs#L1255) but have not yet got it to run, so will post more when that comes through.

**Comment by iancoleman** [comment]
@bochaco I've groked this panic and can submit a pull request with a fix. But first I want to write it up so we can all understand it, however it's a long and complex thing to explain so will take some time. I just wanted to make sure you know **there's no need for you to follow this up**, I'll submit a PR in the (hopefully near) future.

**Comment by iancoleman** [comment]
This is some reasoning why boundary=1 leads to the panic, but also why it is still an acceptable value to use.

The consequence is we need to be serializing the crdt Identity generator `strategy_vec`, but it is [currently not serialized](https://github.com/rust-crdt/rust-crdt/blob/492705097c2ed198a400eb4ed1a941e34bccae1f/src/lseq/ident.rs#L36).

I have done this change locally but immediately came across a new error (which seems unrelated when looking at the error text, but is directly due to the change) so must dig further. Details below.

---

### Background

Firstly some context and background knowledge so we're all on the same page

Crdt data updates are given some numerical id. In the crdts crate this is called `Identifier` with a `path` field ([source](https://github.com/rust-crdt/rust-crdt/blob/492705097c2ed198a400eb4ed1a941e34bccae1f/src/lseq/ident.rs#L11-L19)). The code comment says "A tree identifier uniquely locates an element in an LSeq tree. It represents the path that needs to be taken in order to reach the element. At each level we store the index of the child tree node as well as the id of the site that inserted that node. This resolves conflicts where two sites decide to pick the same child index to allocate a fresh node."

The crate comes with an Identity generator ([source](https://github.com/rust-crdt/rust-crdt/blob/492705097c2ed198a400eb4ed1a941e34bccae1f/src/lseq/ident.rs#L21-L40)). Code comment: "A generator for fresh identifiers. These identifiers represent a path in an exponential tree. At each level of the tree the amount of children doubles. This means that we can store a lot of nodes in a very low height!"

When appending data, the crdts crate automatically generates the next Identifier using the generator function `alloc` ([source](https://github.com/rust-crdt/rust-crdt/blob/492705097c2ed198a400eb4ed1a941e34bccae1f/src/lseq/ident.rs#L86)). In the code comments, "Allocates a new identifier between `p` and `q` such that `p < z < q`."

With initial_base=10 the upper range for the first identifier (z) is 2^10=1024, and is generated such that `0 < z < 1024`.

In the alloc function, we "Descend both identifiers in parallel until we find room to insert an identifier" and end up at [L106](https://github.com/rust-crdt/rust-crdt/blob/492705097c2ed198a400eb4ed1a941e34bccae1f/src/lseq/ident.rs#L106) because "We take this branch if there's room between a and b to insert a new identifier".
L106: `let next_index = self.index_in_range(a.0 + 1, b.0, depth);`

### Expected Behavior

The call to `index_at_range` is where things get interesting.

We are generating a new Identity "in the range [lower, upper)" which is (for the first operation) [0, 1024) and this is done using "the allocation strategy of that depth, [either the strategy called] boundary+ or [the strategy called] boundary-". The strategy is chosen by either a) the previously used strategy for this depth if one has already been used, or b) a random choice between the two strategies (`fn strategy` is the relevant [source](https://github.com/rust-crdt/rust-crdt/blob/492705097c2ed198a400eb4ed1a941e34bccae1f/src/lseq/ident.rs#L254-L264)).

If the strategy is `boundary+` then we calculate the new Identity as `lower + step`

If the strategy is `boundary-` then we calculate the Identity as `upper - step - 1`

Step is (combining a couple of prior lines to simplify)
`step = rng.gen_range(0, std::cmp::min(self.boundary, upper - lower))`
which with boundary=1 comes out as
`step = rng.gen_range(0, std::cmp::min(1, upper - lower))`
which is going to be
`gen_range(0,min(1,1024-0))`
which is
`gen_range(0,1)`
which is `0`

If we happen to randomly end up in `boundary+` strategy we calculate index as `lower+step`, ie `0+0`.

If we happen to randomly end up in `boundary-` strategy we calculate index as `upper-step-1`, ie `1024-0-1` which always puts us right at the top of the range when boundary=1.

Once the strategy is chosen we keep it until this particular depth has no ids left. This is tracked by `strategy_vec` ([source](https://github.com/rust-crdt/rust-crdt/blob/492705097c2ed198a400eb4ed1a941e34bccae1f/src/lseq/ident.rs#L36)), letting us know which strategy was previously use for a given depth (or picks a random strategy if no current strategy exists for the required depth).

So if we happen to pick `boundary+` we start at id 1 and can do 1023 more operations before reaching id 1024 and triggering the next depth. If we happen to pick `boundary-` we start at id 1023 and can do 1 more operation before reaching id 1024 and triggering the next depth. This is one aspect of why boundary=1 is causing rapid progress toward maximum depth 31 where we panic. If boundary was 10 and we start with `boundary-` strategy then the first index is as low as 1014 instead of 1023, so we'd have up to 9 more operations before hitting the next depth.

Since the maximum depth is hard capped at 31 (the original source of the panic) and depth starts at 10, we're allowed about 21 increases in depth before we panic.

If we're unlucky we might get a couple of `boundary-` strategies in a row but any time we hit a `boundary+` strategy we can do many operations at that depth, since every future Identity will be generated from the lower bound of the range and increment from there. This is exactly what we see when we do many appends using the `rust-crdt` or `sn_data_types` repo. As soon as we randomly choose the `boundary+` strategy we can do many many appends without blowing out the depth. But we don't see this with `sn_client` or `sn_api`. Why is that? What's the difference?

### Different Behavior

It turns out with `sn_client` and `sn_api` the `strategy_vec` is always empty, *so we never actually have a previous strategy*. `strategy_vec` is empty because it's not serialized ([source](https://github.com/rust-crdt/rust-crdt/blob/492705097c2ed198a400eb4ed1a941e34bccae1f/src/lseq/ident.rs#L36)), so nodes drop the `strategy_vec` when they serialize chunks, and initialize it empty when they deserialize. This means we end up accidentally choosing a strategy at random *every time*, so we get `boundary-` a lot more often than we normally would and never get the long run that a `boundary+` strategy is supposed to offer.

### Solution

The solution seems easy - change `strategy_vec` so we don't drop it when we serialize. My change to rust-crdt swapped unserializable `BitVec` to serializable `Vec::<bool>` (maybe better to modify BitVec but this was easiest for a quick test). This did not work, `sn_api` throws a new and different error once that change is made.

Unfortunately the new error seems entirely unrelated to the change. It happens when calling `safe seq append` or `safe cat` on the new sequence url.

The error is: `ERROR safe] sn_cli error: [Error] ContentNotFound - No Sequence found at this address`. So I will have to dig deeper on this.

But the good news is boundary=1 should be totally fine so long as the `strategy_vec` is retained.

It would be interesting sometime to document the decision process leading to boundary=1 and tree_base=10.

### Mathematical basis for panics

It becomes clear now why the panic happens after about 40 appends with boundary=1 and tree_base=10.

With tree_base=10 and a max tree_base of 31, we're allowed 31-10=21 depth increases.

There's a `1/boundary` chance of increasing the depth when the strategy is `boundary-`. Since boundary=1 we increase the depth 100% of the time `boundary-` strategy is chosen.

`boundary-` strategy is chosen 50% of the time because the `strategy_vec` is not retained (even though it should be).

So every operation ends up with an overall 50% chance of increasing the depth.

This gives us the chance that `n` operations leads to a total of 21 depth increases (and a panic). eg after 31 append operations there's a 3.54% we'd hit depth 31 and panic. After 47 append operations there's a 80.92% chance we'd hit depth 31 and panic.

| operations (n) | chance of panic (%) |
|------------|---------------------|
| 24         |  0.01               |
| 25         |  0.05               |
| 26         |  0.12               |
| 27         |  0.30               |
| 28         |  0.63               |
| 29         |  1.21               |
| 30         |  2.14               |
| 31         |  3.54               |
| 32         |  5.51               |
| 33         |  8.14               |
| 34         | 11.47               |
| 35         | 15.53               |
| 36         | 20.25               |
| 37         | 25.57               |
| 38         | 31.36               |
| 39         | 37.46               |
| 40         | 43.73               |
| 41         | 50.00               |
| 42         | 56.12               |
| 43         | 61.96               |
| 44         | 67.42               |
| 45         | 72.43               |
| 46         | 76.93               |
| 47         | 80.92               |
| 48         | 84.38               |
| 49         | 87.36               |

**Comment by iancoleman** [comment]
This is fixed in rust-crdt version 5.1.0

From what I can see, this package is used in
* sn_client
* sn_node
* sn_data_types
* sn_messaging

Updating those Cargo.toml files is beyond the scope of this issue so I'll close it.

**Comment by iancoleman** [comment]
I updated crdts to "5.1.0" to confirm this was fixed by that release. It is not fixed, the panic is still there, so the issue will be reopened.

The crates that need updating to crdts = "5.1.0" are

* sn_node
* sn_messaging
* sn_transfers
* sn_data_types
* sn_client

Since this version update needs coordination across all those repos at the same time maybe someone at maidsafe can see to this?

I have a pretty good idea why the panic is still there, but need to write it up (unsurprisingly it's a little complicated so takes some effort to distill into something readable).

No point updating to 5.1.0 until the fix is confirmed. Just wanted to list these crates out for future reference.

**Comment by iancoleman** [comment]
Repeating the original test of appending 50 times to a sequence, I saw 8/20 tests fail using crdts v5.1.0 (whereas with crdts v4.3.0 it was 20/20 failures).

I'll keep digging to see what is causing this intermittent failure. It's the same panic
> maximum depth exceeded

in
> crdts::lseq::ident::IdentGen<A>::alloc

specifically
> Panic occurred in file '/home/ian/code/maidsafe/rust-crdt/src/lseq/ident.rs' at line 214

---

So far no specific hunches but I'm going to be probing sn_data_types seq_crdt.rs especially around `create_append_op` and `apply_data_op`.

My gut feeling is it's to do with creating the append op then separately applying it. I suspect this is possibly messing with the strategy_vec. In `create_append_op` we get a `mut` copy of the data, so the LSeq is being modified by the append, but the mutations are discarded (I think) when we leave that function and possibly not fully retained by the `Op::Insert` (I'm specifically thinking the strategy_vec maybe modified then discarded). Then in `apply_data_op` we get the data LSeq again, and I wonder if it has retained the mutations from `create_append_op` or if data is the non-appended one?

But this might be a red herring, I'll have to dig. Won't get to it for at least a few days so I'm recording my current thinking here to pick up again later.

**Comment by dirvine** [comment]
Excellent @iancoleman  This is one for @bochaco  and @davidrusu  (perhaps @dan-da ) to get involved in. We do need to solve it. Again huge props on another detailed piece of work. We owe you soo soo many beers :-)

**Comment by iancoleman** [comment]
The bug happens because applying an `Op` on the nodes does not update their strategy_vec.

The sequence of events is

* The client creates a sequence and saves it to the nodes. Everything is as expected.

* The client appends data to the sequence. This updates the client version of the LSeq, sometimes increasing the depth of the strategy_vec if needed, and creates a `crdts::Op::Insert` that can be sent to the nodes.

* The client sends *only the op* to the nodes.

* The nodes apply the op, which *should* also increase the strategy_vec depth if needed (with the same new value the client used). But the op alone does not modify the strategy_vec on the nodes.

* The next time the client fetches the sequence (eg to do the next append), the entries are correct, but the strategy_vec is still at the very first value because `crdts::Op::Insert` never updates strategy_vec on the nodes. This means the client generates new strategies far more often than they should, and they rapidly exhaust the available depth.

When the sequence is *created* the whole lseq is sent (including the strategy_vec) ([source](https://github.com/maidsafe/sn_client/blob/320d8eb49bf26a3f880d476d505b9b7670def0b2/src/client/sequence_apis.rs#L176))

When the sequence is *updated* the op is sent, which does not update the node strategy_vec ([source](https://github.com/maidsafe/sn_client/blob/320d8eb49bf26a3f880d476d505b9b7670def0b2/src/client/sequence_apis.rs#L293); also note that L290 is redundant and can be removed since L286 updates the client copy).

Some possible ways to resolve this:

* The lazy way is to send the entire updated data lseq (which will include the updated strategy_vec) to the nodes instead of just the op. This wastes a lot of resources and is not desirable.

* The less lazy way is to include both the Op and the strategy_vec, but this is difficult because the strategy_vec is private within crdts.

* The difficult but correct way is to modify rust-crdt `Op::Insert` to include enough data that all other copies can be updated correctly (including their strategy_vec, not just the entries).

The test I added to rust-crdt that shows this failure (in test/lseq.rs) but needs some addition of `pub` to a couple of fields before it will run...

```
#[test]
fn test_insert_op_remote() {
    let base = 2;
    let boundary = 1;
    let mut site1 = LSeq::new_with_args(0, base, boundary); // the client

    // write the entire existing lseq to site2
    let mut site2 = site1.clone(); // a node

    // site1 appends data and sends only the insert op to site2
    for _ in 0..10000 {
        let op = site1.append('b');
        site1.apply(op.clone());
        site2.apply(op.clone());
    }   

    assert_eq!(site1.gen.strategy_vec, site2.gen.strategy_vec);
}
// output is:
//
// 'assertion failed: `(left == right)`
//  left: `[true, false, false, true, false, false, false, false, false, false, true, true]`,
// right: `[]`'
//
// The client updates their strategy_vec, but the node does not.
// If the client fetches from the node, they will have an empty strategy_vec no
// matter how many appends have already happened.
```

I will try to work out a PR for rust-crdt to fix this, although it feels possibly stylistically open-ended, so maybe an issue is more appropriate to discuss it further. @davidrusu does this make sense to you?

**Comment by bochaco** [comment]
Thanks a lot @iancoleman for the whole write up, it seems you also found the bug and potential solution. I didn't make it to test this yet today, was just trying to catch up with it, I'll try to reproduce it tomorrow. 

How about generating the strategy in a deterministic way, so all replicas choose the same one for each level and no need to sync-up and/or send it with the op data?
I don't remember if this was covered in the LSeq paper, I'd be surprised if it's not, but it sounds like we should either be able to set one strategy for all levels of the tree (either + or -), or if it's random then it'll have to be deterministically random, e.g. boundary+ for even depth and boundary- for odd levels.

**Comment by bochaco** [comment]
I took a quick look again at the LSeq paper (https://hal.archives-ouvertes.fr/hal-00921633/document), and in section 3 it mentions about the pros/cons of using boundary+/-, I think for our public Sequence we can simply stick to boundary+ for all depths, and perhaps random but deterministically go for boundary+/- for private Sequence as the user may be inserting entries and not just appending to the end as with public Sequence.

**Comment by bochaco** [comment]
I also reviewed some old code I worked on for an alternative implementation of LSeq, I see I was using some idea of deterministic strategies back then https://github.com/bochaco/rust-crdt/blob/lseq-crdt/src/lseq/lseq.rs#L16, I'll keep refreshing my memory and try to reproduce it tomorrow

**Comment by davidrusu** [comment]
Thanks for doing this research @iancoleman.

If I remember correctly, each strategy vec was to be used only by a single actor to avoid any colliding identifiers, perhaps we can have a strategy vec per actor stored in a map alongside the LSEQ.

I'm also with @bochaco on deterministicly generating identifiers, nodes should be able to verify that a client generated this identifier in a kosher way otherwise dishonest nodes can craft identifiers to force the depth of the tree to increase and trigger these same panics.

**Comment by davidrusu** [comment]
All this investigation made me think that we may have more luck with a purpose built CRDT for modelling the append only structure we are after here.

I wrote up the design for a `Chain` CRDT here:
https://github.com/rust-crdt/rust-crdt/pull/103

I think the structure is quite simple and handles concurrent appends in a reasonably intuitive way. I'd be curious to hear any thoughts (on the PR as to not distract from this discussion here).

**Comment by bochaco** [comment]
I'll be taking a look at that @davidrusu and comments there

**Comment by iancoleman** [comment]
The suggestions sound like there's plenty of possible solutions, I'm sure one is going to be especially good, it's just a matter of deciding the most appropriate one. My understanding of crdts is relatively shallow, so all I feel I can contribute here is 'if the solution prevents panic then I'm happy'.

> I took a quick look again at the LSeq paper (https://hal.archives-ouvertes.fr/hal-00921633/document), and in section 3 it mentions about the pros/cons of using boundary+/-

Thanks for the link to the LSeq paper @bochaco I hadn't seen it before. Might be worth adding it to the rust-crdt readme. I was especially interested in the base/boundary values they used for their Wikipedia experiment (initial_base=4, boundary=10).

> If I remember correctly, each strategy vec was to be used only by a single actor to avoid any colliding identifiers, perhaps we can have a strategy vec per actor stored in a map alongside the LSEQ.

This sounds like it would work. I wonder though, if we have a large enough boundary then it shouldn't matter if everyone is using the same strategy_vec, right?

I feel like some docs helping app developers choose a suitable initial_base and boundary value for their particular usage would be very helpful to have at some stage.

> I think for our public Sequence we can simply stick to boundary+ for all depths

I tried a quick hack to try always using boundary+ strategy and it worked fine (as expected) for many appends. I got to about 150 appends before performance became quite slow and I stopped. But performance is beyond the scope of this issue, so let's leave it there...

[rust-crdt/src/lseq/ident.rs#L256](https://github.com/rust-crdt/rust-crdt/blob/5.1.0/src/lseq/ident.rs#L256)

```diff
-let new_strategy = rand::thread_rng().gen();
+let new_strategy = true; // always use boundary+ strategy
```

Thanks for the comments @davidrusu and @bochaco

**Comment by davidrusu** [comment]
@iancoleman I've reworked the LSeq CRDT so that it can't panic, we no longer use strategy vecs instead we choose the midpoint between adjacent identifiers. We've also switched to BigRationals for the identifier rather than an exponential tree. This turns out to be very space efficient in the append only use-case.

The new list crdt is fairly straightforward to read through as well. It punts the majority of the complexity to the BigRational internals.
* https://github.com/rust-crdt/rust-crdt/blob/master/src/list.rs
* https://github.com/maidsafe/sn_data_types/pull/335
